src/metabase/models/pulse.clj

```diff
@@ -450,54 +450,49 @@
-      ;; 2. NOT in channels, in db-channels = DELETE
-      (and (nil? channel) existing-channel) (t2/delete! PulseChannel :id (:id existing-channel))
-      ;; 3. in channels, in db-channels = UPDATE
-      (and channel existing-channel)        (pulse-channel/update-pulse-channel! channel)
-      ;; 4. NOT in channels, NOT in db-channels = NO-OP
-      :else                                 nil)))
-
 (mu/defn update-notification-channels!
   "Update the PulseChannels for a given `notification-or-id`. `channels` should be a definitive collection of *all* of
   the channels for the Notification.
 
-   * If a channel in the list has no existing `PulseChannel` object, one will be created.
+    * If a channel in the list has no existing `PulseChannel` object, one will be created.
 
-   * If an existing `PulseChannel` has no corresponding entry in `channels`, it will be deleted.
+    * If an existing `PulseChannel` has no corresponding entry in `channels`, it will be deleted.
 
-   * All previously existing channels will be updated with their most recent information."
+    * All previously existing channels will be updated with their most recent information."
   [notification-or-id channels :- [:sequential :map]]
-  (let [new-channels   (group-by (comp keyword :channel_type) channels)
-        old-channels   (group-by (comp keyword :channel_type) (t2/select PulseChannel
-                                                                         :pulse_id (u/the-id notification-or-id)))
-        handle-channel #(create-update-delete-channel! (u/the-id notification-or-id)
-                                                       (first (get new-channels %))
-                                                       (first (get old-channels %)))]
-    (assert (zero? (count (get new-channels nil)))
-            "Cannot have channels without a :channel_type attribute")
-    ;; don't automatically archive this Pulse if we end up deleting its last PulseChannel -- we're probably replacing
-    ;; it with a new one immediately thereafter.
+  (let [existing-channels                 (t2/select :model/PulseChannel :pulse_id (u/the-id notification-or-id))
+        channel-type+id->existing-channel (update-vals (group-by (juxt :channel_type :channel_id) channels) first)
+        channels                          (map-indexed
+                                           (fn [idx channel]
+                                             (assoc channel
+                                                    :channel_type   (keyword (:channel_type channel))
+                                                    :schedule_type  (keyword (:schedule_type channel))
+                                                    :schedule_frame (keyword (:schedule_frame channel))
+                                                    :pulse_id       (u/the-id notification-or-id)
+                                                    ;; channel from FE might not have an id, so we need to find the id
```

Comment:
is this current behavior? I wonder if it makes sense to change that that and then this function will be much simpler?

Reply:
turns out we don't need it at all :) removed